### PR TITLE
Support Payment Allocation with different currencies

### PR DESCRIPTION
**Pull request by TheBestPessimist**
https://github.com/metasfresh/metasfresh/issues/9897

### CODE REVIEW

File: backend/de.metas.adempiere.adempiere/client/src/main/java/org/compiere/apps/search/FindHelper.java [hunk 1]
```diff
@@ -35,16 +34,13 @@
 import de.metas.adempiere.util.Permutation;
 import de.metas.util.Check;
 
-/**
- * 
- * @author tsa
- * 
- */
+import javax.annotation.Nullable;
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/client/src/main/java/org/compiere/apps/search/FindHelper.java:37]: pls avoid changing files while are not related to the issue.
makes reviewing even more harder.

File: backend/de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/payment/paymentallocation/service/PaymentAllocationBuilder.java [hunk 2]
```diff
@@ -40,6 +67,9 @@
 		DO_NOTHING, WRITE_OFF, DISCOUNT
 	}
 
+	private final MoneyService moneyService = SpringContextHolder.instance.getBean(MoneyService.class);
+	private final ICurrencyBL currencyBL = Services.get(ICurrencyBL.class);
```
Comment (teosarca) [inline backend/de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/payment/paymentallocation/service/PaymentAllocationBuilder.java:71]: pls move those near the other services declared in this class... see line 78

File: backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/InvoiceRow.java [hunk 3]
```diff
@@ -533,7 +535,7 @@ else if (invoiceRow.getC_Order_ID() > 0)
 		final CurrencyId currencyId = currenciesRepo.getByCurrencyCode(invoiceRow.getCurrencyISOCode()).getId();
 
 		return PayableDocument.builder()
-				.orgId(invoiceRow.getOrgId())
+				.clientAndOrgId(ClientAndOrgId.ofClientAndOrg(ClientId.METASFRESH, invoiceRow.getOrgId()))
```
Comment (teosarca) [inline backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/InvoiceRow.java:538]: avoid avoid hardcoding ClientId

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/payment_allocation/process/PaymentsViewAllocateCommand.java [hunk 4]
```diff
@@ -210,9 +209,17 @@ static PayableDocument toPayableDocument(
...
 				.invoiceProcessingFeeCalculation(invoiceProcessingFeeCalculation)
+				.date(row.getDateInvoiced())
+				.clientAndOrgId(row.getClientAndOrgId())
+				.currencyConversionTypeId(row.getCurrencyConversionTypeId())
 				.build();
 	}
 
+	private static boolean hasServiceFeesToPay(final @NonNull InvoiceRow row, final CurrencyCode currencyCode)
+	{
+		return row.getServiceFeeAmt() != null && !Amount.zero(currencyCode).equals(row.getServiceFeeAmt());
```
Comment (teosarca) [inline backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/payment_allocation/process/PaymentsViewAllocateCommand.java:220]: why not using Amount.isZero instead of that weird "equals" comparation....

File: backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/PaymentRow.java [hunk 5]
```diff
@@ -242,14 +244,15 @@ public IPaymentDocument copyAsPaymentDocument()
 		final CurrencyId currencyId = currenciesRepo.getByCurrencyCode(paymentRow.getCurrencyISOCode()).getId();
 
 		return PaymentDocument.builder()
-				.orgId(paymentRow.getOrgId())
+				.clientAndOrgId(ClientAndOrgId.ofClientAndOrg(ClientId.METASFRESH, paymentRow.getOrgId()))
```
Comment (teosarca) [inline backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/PaymentRow.java:247]: avoid avoid hardcoding ClientId.
ClientId.METASFRESH is manly for testing!

File: backend/de.metas.adempiere.adempiere/client/src/main/java/org/compiere/apps/search/FindHelper.java (hunk 1 above)
Comment (TheBestPessimist) [reply backend/de.metas.adempiere.adempiere/client/src/main/java/org/compiere/apps/search/FindHelper.java:37]: what about the boyscout rule?

File: backend/de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/payment/paymentallocation/service/PaymentAllocationBuilder.java (hunk 2 above)
Comment (TheBestPessimist) [reply backend/de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/payment/paymentallocation/service/PaymentAllocationBuilder.java:71]: done

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/payment_allocation/process/PaymentsViewAllocateCommand.java (hunk 4 above)
Comment (TheBestPessimist) [reply backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/payment_allocation/process/PaymentsViewAllocateCommand.java:220]: done

File: backend/de.metas.adempiere.adempiere/client/src/main/java/org/compiere/apps/search/FindHelper.java (hunk 1 above)
Comment (TheBestPessimist) [reply backend/de.metas.adempiere.adempiere/client/src/main/java/org/compiere/apps/search/FindHelper.java:37]: plus: reviewing these side-touched files should be relatively easy, because most of the changes are: 
- nullability annotations
- removing parameter names from comments if they dont have javadoc
- cleanup of final/static when lombok already ads that
- fixing typos

File: backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/InvoiceRow.java (hunk 3 above)
Comment (TheBestPessimist) [reply backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/InvoiceRow.java:538]: Problem: i don't have a client id here and i believe org is null as well as it is never set to the builder. 
I guess this never failed because it was never used.
Is there a point in investing any time in fixing this, since this belongs to swing?

Can i use `Env.getClientId()` and `Env.getOrgId()` to get some generic values?

File: backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/PaymentRow.java (hunk 5 above)
Comment (TheBestPessimist) [reply backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/PaymentRow.java:247]: Problem: i don't have a client id here and i believe org is null as well as it is never set to the builder. 
I guess this never failed because it was never used.
Is there a point in investing any time in fixing this, since this belongs to swing?

Can i use `Env.getClientId()` and `Env.getOrgId()` to get some generic values?

File: backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/InvoiceRow.java (hunk 3 above)
Comment (teosarca) [reply backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/InvoiceRow.java]: agree. don't invest time in swing.

File: backend/de.metas.adempiere.adempiere/client/src/main/java/org/compiere/apps/search/FindHelper.java (hunk 1 above)
Comment (teosarca) [reply backend/de.metas.adempiere.adempiere/client/src/main/java/org/compiere/apps/search/FindHelper.java:37]: boy scout rule says to clear the place near you, not the place in the next forest :)
here, to be more precise, I wasn't sure if there were some changes related to this task or just some cleanup.
thanks for explaining.

File: backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/PaymentRow.java (hunk 5 above)
Comment (teosarca) [reply backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/PaymentRow.java]: agree. don't invest time in swing.

Comment (TheBestPessimist) [reply backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/PaymentRow.java]: fixed

File: backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/InvoiceRow.java (hunk 3 above)
Comment (TheBestPessimist) [reply backend/de.metas.banking/de.metas.banking.swingui/src/main/java/de/metas/banking/payment/paymentallocation/model/InvoiceRow.java]: fixed

File: backend/de.metas.banking/de.metas.banking.base/src/main/sql/postgresql/system/5566700_sys_gh9897_UpdateFunction_GetOpenPayments.sql [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  AND p.IsAllocated='N'
+  AND p.Processed='Y'
+  AND p.C_Charge_ID IS NULL
+  AND ($3 = 'Y' OR p.C_Currency_ID=$2)
+  AND ($4 IS NULL OR $4=0 or p.AD_Org_ID=$4)
+ORDER BY
+    p.DateTrx, p.DocumentNo
+    ;
+$BODY$
+    LANGUAGE sql VOLATILE
```
Comment (metas-ts) [inline backend/de.metas.banking/de.metas.banking.base/src/main/sql/postgresql/system/5566700_sys_gh9897_UpdateFunction_GetOpenPayments.sql:88]: can proooobably be stable

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.banking/de.metas.banking.base/src/main/sql/postgresql/system/5566700_sys_gh9897_UpdateFunction_GetOpenPayments.sql (hunk 6 above)
Comment (TheBestPessimist) [reply backend/de.metas.banking/de.metas.banking.base/src/main/sql/postgresql/system/5566700_sys_gh9897_UpdateFunction_GetOpenPayments.sql]: done

**Comment by TheBestPessimist** [comment]
I agree with the stable proposition.

**Review by metas-ts — APPROVED** [review]
